spectrocloud/resource_application_profile.go

```diff
@@ -0,0 +1,546 @@
+		}
+		break
+	}*/
+
+	tier := &models.V1AppTierEntity{
+		Name:             ptr.StringPtr(pName),
+		Version:          pVersion,
+		SourceAppTierUID: source_app_tier,
+		RegistryUID:      pRegistryUID,
+		//UID:         pUID,
+		Type: pType,
+		// UI strips a single newline, so we should do the same
+		Values: strings.TrimSpace(p["values"].(string)),
+	}
+
+	manifests := make([]*models.V1ManifestInputEntity, 0)
+	if len(p["manifest"].([]interface{})) > 0 {
+		for _, manifest := range p["manifest"].([]interface{}) {
+			m := manifest.(map[string]interface{})
+			manifests = append(manifests, &models.V1ManifestInputEntity{
+				Content: strings.TrimSpace(m["content"].(string)),
+				Name:    m["name"].(string),
+			})
+		}
+	}
+	tier.Manifests = manifests
+
+	return tier, nil
+}
+
+// get update create delete separately based on previous version.
+func toApplicationTiersUpdate(d *schema.ResourceData, c *client.V1Client) ([]*models.V1AppTierEntity, map[string]*models.V1AppTierUpdateEntity, []string, error) {
+	previousTiers, err := c.GetApplicationProfileTiers(d.Id())
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	previousTiersMap := map[string]*models.V1AppTier{}
+	for _, tier := range previousTiers {
+		previousTiersMap[tier.Metadata.Name] = tier
+	}
+
+	var createTiers []*models.V1AppTierEntity
+	updateTiersMap := map[string]*models.V1AppTierUpdateEntity{}
+	updateTiersMapId := map[string]*models.V1AppTierUpdateEntity{}
+	var deleteTiers []string
+
+	createTiersMap := map[string]*models.V1AppTierEntity{}
+	tiers := make([]*models.V1AppTierEntity, 0)
+	for _, tier := range d.Get("pack").([]interface{}) {
+
+		if _, found := previousTiersMap[tier.(map[string]interface{})["name"].(string)]; found {
+			t := toApplicationProfilePackUpdate(tier)
+			updateTiersMap[t.Name] = t
+		} else {
+			if t, e := toApplicationProfilePackCreate(tier); e != nil {
+				return nil, nil, nil, e
+			} else {
+				createTiers = append(createTiers, t)
+				tiers = append(tiers, t)
+				createTiersMap[*t.Name] = t
+			}
+		}
+	}
+
+	for _, tier := range previousTiers {
+		_, create := createTiersMap[tier.Metadata.Name]
+		_, update := updateTiersMap[tier.Metadata.Name]
+		if !create && !update {
+			deleteTiers = append(deleteTiers, tier.Metadata.UID)
+		}
+		if update {
+			updateTiersMapId[tier.Metadata.UID] = updateTiersMap[tier.Metadata.Name]
+		}
+	}
+
+	return createTiers, updateTiersMapId, deleteTiers, nil
+
+}
+
+func toApplicationProfilePatch(d *schema.ResourceData) (*models.V1AppProfileMetaEntity, error) {
+	description := ""
+	if d.Get("description") != nil {
+		description = d.Get("description").(string)
+	}
+	metadata := &models.V1AppProfileMetaEntity{
+		Metadata: &models.V1AppProfileMetaUpdateEntity{
+			//TODO name change?: Name: d.Get("name").(string),
+			Annotations: map[string]string{
+				"description": description,
+			},
+			Labels: toTags(d),
+		},
+		/*TODO: check profile version: Spec: &models.V1ClusterProfileSpecEntity{
```

Comment:
is this a real TODO?

Reply:
This will be needed later when we support profile create via clone. Real TODO.